price-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+#![no_std]
+
+use crate::redeem_token::{ACCEPTED_TOKEN_REDEEM_NONCE, LAUNCHED_TOKEN_REDEEM_NONCE};
+
+elrond_wasm::imports!();
+
+pub mod common_storage;
+pub mod create_pool;
+pub mod redeem_token;
+
+const INVALID_PAYMENT_ERR_MSG: &[u8] = b"Invalid payment token";
+
+#[elrond_wasm::contract]
+pub trait PriceDiscovery:
+    common_storage::CommonStorageModule
+    + create_pool::CreatePoolModule
+    + redeem_token::RedeemTokenModule
+{
+    #[init]
+    fn init(
+        &self,
+        dex_sc_address: ManagedAddress,
+        launched_token_id: TokenIdentifier,
+        accepted_token_id: TokenIdentifier,
+        start_epoch: u64,
+        end_epoch: u64,
+    ) -> SCResult<()> {
+        require!(
+            self.blockchain().is_smart_contract(&dex_sc_address),
+            "Invalid DEX SC address"
+        );
+        require!(
+            launched_token_id.is_valid_esdt_identifier(),
+            "Invalid launched token ID"
+        );
+        require!(
+            accepted_token_id.is_egld() || accepted_token_id.is_valid_esdt_identifier(),
+            "Invalid payment token ID"
+        );
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        require!(
+            current_epoch < start_epoch,
+            "Start epoch cannot be in the past"
+        );
+        require!(current_epoch < end_epoch, "End epoch cannot be in the past");
+        require!(
+            start_epoch < end_epoch,
+            "Start epoch must be before end epoch"
+        );
+
+        self.dex_sc_address().set(&dex_sc_address);
+        self.launched_token_id().set(&launched_token_id);
+        self.accepted_token_id().set(&accepted_token_id);
+        self.start_epoch().set(&start_epoch);
+        self.end_epoch().set(&end_epoch);
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[payable("*")]
+    #[endpoint(depositInitialTokens)]
+    fn deposit_initial_tokens(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let start_epoch = self.start_epoch().get();
+        require!(
+            current_epoch < start_epoch,
+            "May only deposit before start epoch"
+        );
+
+        let launched_token_id = self.launched_token_id().get();
+        require!(payment_token == launched_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, LAUNCHED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn deposit(
```

Comment:
I think only one deposit is enough. And according to tokenID you go for the tokenA or tokenB nonce (initial token or launchpad token).

Reply:
Done.